rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java

```diff
@@ -33,12 +33,21 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+/* Important invariant that must always hold:
+   A key cannot be in the committedCache and uncommittedCache cache at the same time.
+   This is because, for performance reasons it seams, get() queries the committedCache BEFORE the
+   uncommittedCache.
+
+ */
 public class DataSourceWithCache implements KeyValueDataSource {
 
     private static final Logger logger = LoggerFactory.getLogger("datasourcewithcache");
 
     private final int cacheSize;
+    private final boolean preventWritesToBase;
     private final KeyValueDataSource base;
+    // When preventWritesToBase=true, uncommittedCache will forever contain
+    // the changes, and committedCache will still work as a cache.
```

Comment:
Wouldn't it be better to not use uncommittedCache at all in readonly mode? At least, we would be consuming less memory.

On a quick look it would just be a matter of updating org.ethereum.datasource.DataSourceWithCache#putKeyValue to update committedCache when readonly.

Reply:
I cannot find this line in the final code.

Reply:
Never mind, I got it wrong back then. We need the `uncommittedCache` as the actual memory store and `committedCache` as well because it can contain data from the snapshot.